### PR TITLE
feat: adaptable min difficulty check

### PR DESCRIPTION
**Pull request by SWvheerden**
Description
---
This includes an adaptable min difficulty check for incoming blocks. 

Motivation and Context
---
See issue: #5620 for the reasoning behind why this needs to include an adaptable min difficulty check.

### CODE REVIEW

File: base_layer/core/src/proof_of_work/difficulty.rs [hunk 1]
```diff
@@ -83,6 +83,18 @@ impl Difficulty {
...
+        if other == 0 {
+            return None;
+        }
+        let result = Difficulty(self.0 / other);
+        if result.0 < MIN_DIFFICULTY {
+            None
+        } else {
+            Some(result)
+        }
+    }
```
Comment (brianp) [inline base_layer/core/src/proof_of_work/difficulty.rs]: There is a `checked_div` function for u64 already. It does divide by 0 checking. Is there a reason not to use it?
https://doc.rust-lang.org/stable/std/primitive.u64.html#method.checked_div

Additionally, this function is doing more than a standard `checked_*` function. We're comparing it against our own min difficulty. So we could go for a name that doesn't match a pre-existing usage pattern.

```suggestion
    pub fn min_difficulty_checked_div(&self, other: u64) -> Option<Difficulty> {
        return match self.0.checked_div(other) {
            None => None,
            Some(n) => {
                n < MIN_DIFFICULTY ? None : Some(Difficulty(n))
            }
        }
    }
```

Comment (brianp) [reply base_layer/core/src/proof_of_work/difficulty.rs]: Additionally, the `MIN_DIFFICULTY` constant is well... Constant. Shouldn't this be compared against the current minimum proof of work difficulty?

Is there a chance someone could use this to always achieve the minimum difficulty of 1? n could be the same as the constant which would result in a min difficulty lower than the pow difficulty.

**Review by brianp — CHANGES_REQUESTED** [review]
Some suggestions for the `checked_div_u64` functions. I probably need a clarification on the way we're using the min difficulty we're calculating because it looks like it could be possible to keep the value at 1.

File: base_layer/core/src/proof_of_work/difficulty.rs (hunk 1 above)
Comment (SWvheerden) [reply base_layer/core/src/proof_of_work/difficulty.rs]: Min_difficulty is a constant, and that why we need to keep checking it against that. And the context here keeps it to difficulty, and does not go towards having a min_chain difficulty that dependant on previous timestamps etc

**Review by brianp — APPROVED** [review]
LGTM, utAck

**Pull request merged**